File: challenges/xlang/ml/fannkuch.ml
let fannkuch n =
  let perm = Array.make n 0 in
  let perm1 = Array.init n (fun i -> i) in
  let count = Array.make n 0 in
  let max_flips = ref 0 and checksum = ref 0 and perm_count = ref 0 in
  let r = ref n in
  let result = ref (0, 0) in
  let finished = ref false in
  while not !finished do
    while !r <> 1 do count.(!r - 1) <- !r; decr r done;
    Array.blit perm1 0 perm 0 n;
    let flips = ref 0 in
    let continue_flip = ref true in
    while !continue_flip do
      let k = perm.(0) in
      if k = 0 then continue_flip := false
      else begin
        let k2 = (k + 1) lsr 1 in
        for i = 0 to k2 - 1 do
          let t = perm.(i) in perm.(i) <- perm.(k - i); perm.(k - i) <- t
        done;
        incr flips
      end
    done;
    if !flips > !max_flips then max_flips := !flips;
    if !perm_count mod 2 = 0 then checksum := !checksum + !flips else checksum := !checksum - !flips;
    let advancing = ref true in
    while !advancing && not !finished do
      if !r = n then begin result := (!checksum, !max_flips); finished := true end
      else begin
        let perm0 = perm1.(0) in
        let i = ref 0 in
        while !i < !r do perm1.(!i) <- perm1.(!i + 1); incr i done;
        perm1.(!r) <- perm0;
        count.(!r) <- count.(!r) - 1;
        if count.(!r) > 0 then advancing := false else incr r
      end
    done;
    incr perm_count
  done;
  !result

let () =
  let n = if Array.length Sys.argv > 1 then int_of_string Sys.argv.(1) else 7 in
  let (checksum, max_flips) = fannkuch n in
  Printf.printf "%d\nPfannkuchen(%d) = %d\n" checksum n max_flips

File: challenges/xlang/rs/fannkuch.rs
fn fannkuch(n: usize) -> (i32, i32) {
    let mut perm = vec![0usize; n];
    let mut perm1: Vec<usize> = (0..n).collect();
    let mut count = vec![0usize; n];
    let (mut max_flips, mut checksum, mut perm_count) = (0i32, 0i32, 0i64);
    let mut r = n;
    loop {
        while r != 1 { count[r - 1] = r; r -= 1; }
        perm.copy_from_slice(&perm1);
        let mut flips = 0i32;
        loop {
            let k = perm[0];
            if k == 0 { break; }
            let k2 = (k + 1) >> 1;
            for i in 0..k2 { perm.swap(i, k - i); }
            flips += 1;
        }
        if flips > max_flips { max_flips = flips; }
        checksum += if perm_count % 2 == 0 { flips } else { -flips };
        loop {
            if r == n { return (checksum, max_flips); }
            let perm0 = perm1[0];
            let mut i = 0;
            while i < r { perm1[i] = perm1[i + 1]; i += 1; }
            perm1[r] = perm0;
            count[r] -= 1;
            if count[r] > 0 { break; }
            r += 1;
        }
        perm_count += 1;
    }
}

fn main() {
    let n: usize = std::env::args().nth(1).and_then(|a| a.parse().ok()).unwrap_or(7);
    let (checksum, max_flips) = fannkuch(n);
    println!("{}\nPfannkuchen({}) = {}", checksum, n, max_flips);
}
